fplll/pruner/pruner.h

```diff
@@ -462,49 +673,109 @@ template <class FT> class Pruner
+  FT single_enum_cost_evec(/*i*/ const evec &b, vector<double> *detailed_cost = nullptr);
+  FT single_enum_cost_lower(/*i*/ const vec &b, vector<double> *detailed_cost = nullptr);
+  FT single_enum_cost_upper(/*i*/ const vec &b, vector<double> *detailed_cost = nullptr);
 
   /**
      @brief Compute the success probability for SVP/CVP of a single enumeration
      @param b pruning bounds
      @return success probability
+
+     Function svp_probability() is the main interface.
+     Function svp_probability_lower() is an auxiliary function to compute the
+                                      approx. lower bound of succ. probability
+     Function svp_probability_upper() is an auxiliary function to compute the
+                                      upper bound of succ. probability
+     The svp_probability_evec() is the underlying backend for above.
   */
+  FT svp_probability(/*i*/ const vec &b);
+  FT svp_probability_evec(/*i*/ const evec &b);
+  FT svp_probability_lower(/*i*/ const vec &b);
+  FT svp_probability_upper(/*i*/ const vec &b);
 
-  FT svp_probability(/*i*/ const evec &b);
   /**
      @brief Compute the expected number of solution of a single of a single enumeration
      @param b pruning bounds
      @return expected number of solutions
+
+     The naming conventions is the same as those in svp_probability_*()
   */
-  FT expected_solutions(/*i*/ const evec &b);
+  FT expected_solutions(/*i*/ const vec &b);
+  FT expected_solutions_evec(/*i*/ const evec &b);
+  FT expected_solutions_lower(/*i*/ const vec &b);
+  FT expected_solutions_upper(/*i*/ const vec &b);
 
   /**
      @brief links to either svp_probability() or expected_solution() depending on metric
      @param b pruning bounds
      @return expected number of solutions or success probability
   */
-  FT measure_metric(/*i*/ const evec &b);
+  FT measure_metric(/*i*/ const vec &b);
 
   /**
-     Compute the cost of r enumeration and (r-1) preprocessing,
-     where r is the required number of retrials to reach target/target_solution
-     @brief cost of repeating enumeration and preprocessing
+     @brief Compute the target function to be optimized. This could be the
+     cost of repeating enumeration and preprocessing until reaching target
+     or the cost of a single enumeration.
+
+     Note this function does not do optimization. Depending on the
+     optimization goal, it will return the cost based on different
+     models/modes. There are two modes depending on the target.
+
+     (1) if target == -1, it will compute the cost of r * single_enum_cost()
+         plus (r-1) * preprocessing where r is the expected number of
+         trials.
+
+     (2) if target != -1 (target must also be given), it will compute
+         the cost of a single_enum_cost(). Note in such case the input
+         target prob is fixed. Whether we could achieve this input
+         target prob will be determined in the optimization procedure.
+         See also below.
```

Comment:
I think this documentation needs updating (?)

Reply:
yes, updated now :)